src/fam.rs

```diff
@@ -0,0 +1,623 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+
+use std::mem::{self, size_of};
+
+/// Errors associated with the [`FamStructWrapper`](struct.FamStructWrapper.html) struct.
+#[derive(Clone, Debug)]
+pub enum Error {
+    /// The max size has been exceeded
+    SizeLimitExceeded,
+}
+
+/// Trait for accessing some properties of C defined FAM structures.
+///
+/// In C 99 an array of unknown size may appear within a struct definition as the last member
+/// (as long as there is at least one other named member).
+/// This is known as a flexible array member (FAM).
+/// Pre C99, the same behavior could be achieved using zero length arrays.
+///
+/// Flexible Array Members are the go-to choice for working with large amounts of data
+/// prefixed by header values.
+///
+/// For example the kvm API has many structures that are very similar to the following
+/// `MockFamStruct` structure:
+///
+/// # Example
+///
+/// ```
+/// #[repr(C)]
+/// #[derive(Clone, Default)]
+/// struct MockFamStruct {
+///     pub len: u32,
+///     pub padding: u32,
+///     pub entries: [u32; 0],
+/// }
+/// ```
+///
+#[allow(clippy::len_without_is_empty)]
+pub unsafe trait FamStruct {
```

Comment:
Looks good, we should now think what the contract is such that implementing `FamStruct` is actually safe, so we can add it to the documentation. So far, likely mentions are the implementer and `Entry` should essentially be POD (plain old data) structs, and `size_of(Entry)` should be a multiple of `align_of(Entry)`.

Reply:
I didn't write anything related to: `size_of(Entry) should be a multiple of align_of(Entry)`

We'll sync offline.